src/capi/impl/capi_sframe.cpp

```diff
@@ -170,18 +170,171 @@ EXPORT tc_sframe* tc_sframe_join_on_single_column(
+    }
+    config.insert({"column_type_hints", ft_column_type_hints});
+  }
+
+  // na_values: flex_list<flexible_type>
+  it = params->value.find("na_values");
+  if (it != params->value.end()) {
+    std::vector<turi::flexible_type> ft_na_values = turi::variant_get_ref<turi::flexible_type>(it->second).to<turi::flex_list>();
+    for (auto iter = ft_na_values.begin(); 
+      iter != ft_na_values.end(); ++iter) {
+      if ((*iter).get_type() != turi::flex_type_enum::STRING) {
+        throw std::string("Invalid input to na_values optional parameter: requires a flex_list of strings");
+        return NULL;
+      }
+    }
+    config.insert({"na_values", ft_na_values});
+  }
+
+  // line_terminator: string
+  it = params->value.find("line_terminator");
+  if (it != params->value.end()) {
+    std::string line_terminator = turi::variant_get_ref<turi::flexible_type>(it->second).to<turi::flex_string>();
+    config.insert({"line_terminator", line_terminator});
+  }
+
+  // usecols: flex_list<string>
+  it = params->value.find("usecols");
+  if (it != params->value.end()) {
+    std::vector<turi::flexible_type> ft_usecols = turi::variant_get_ref<turi::flexible_type>(it->second).to<turi::flex_list>();
+    for (auto iter = ft_usecols.begin() ; iter != ft_usecols.end(); ++iter) {
+      if ((*iter).get_type() != turi::flex_type_enum::STRING) {
+        throw std::string("Invalid input to usecols optional parameter: requires a flex_list of strings");
+        return NULL;
+      }
+    }
+    config.insert({"usecols", ft_usecols});
+  }
+
+  // nrows: int
+  it = params->value.find("nrows");
+  if (it != params->value.end()) {
+    int64_t nrows = turi::variant_get_ref<turi::flexible_type>(it->second).to<turi::flex_int>();
+    config.insert({"nrows", nrows});
+  }
+
+  // skiprows: int
+  it = params->value.find("skiprows");
+  if (it != params->value.end()) {
+    int64_t skiprows = turi::variant_get_ref<turi::flexible_type>(it->second).to<turi::flex_int>();
+    config.insert({"skiprows", skiprows});
+  }
+
+  // verbose: int
+  it = params->value.find("verbose");
+  if (it != params->value.end()) {
+    int64_t verbose = turi::variant_get_ref<turi::flexible_type>(it->second).to<turi::flex_int>();
+    config.insert({"verbose", verbose});
+  }
+  
   ret->value.construct_from_csvs(url, config, column_type_hints);
+  
   return ret;
 
   ERROR_HANDLE_END(error, NULL);
 }
 
+EXPORT tc_sframe* tc_sframe_read_json_lines(const char *url, tc_error **error) {
+  ERROR_HANDLE_START();
+
+  tc_sframe* ret = new_tc_sframe();
+  turi::csv_parsing_config_map config;
+  turi::str_flex_type_map column_type_hints;
+
+  config.insert({"header", 0});
+
+  ret->value.construct_from_csvs(url, config, column_type_hints);
+  if (ret->value.num_columns() != 1) {
+    throw std::string("Input JSON not of expected format");
+    return NULL;
+  }
+  if (ret->value["X1"].dtype() == turi::flex_type_enum::DICT) {
+    return (new_tc_sframe(ret->value.unpack("X1","")));
```

Comment:
Should we be unpacking the resulting dict in the CAPI layer? Or is this logic that should go deeper into `construct_from_csvs` to be shared with Python, etc.? I don't think we necessarily want to diverge behavior here between CAPI and other uses of our API.

Reply:
This is true -- the issue is that this is currently done in python, so this is just the change with minimal surface area.  Let's deal with that in a separate PR.

Reply:
Yes, I agree -- this logic should not be in python or here.